Migrate data flow section (#72)
cossacklabs/product-docs: acra/acra-in-depth/data-flow/_index.md

---
title: Data flow
bookCollapseSection: true
---

## Data flow

### Acra components

> - [_AcraWriter_](/pages/documentation-acra/#acraconnector-and-acrawriter) — a client-side library that integrates into the app's workflow either through ORM or directly and provides the means for encryption of the sensitive data by generating AcraStructs.

Reviewer: Is `/pages/` some kind of redirection? Because we only have `acra` and `themis` at the root of the repo.
Also, maybe `apps` would be better than `app's` since second one is generally short form of `app is`.

Author: nice catch, I missed this link

  ![](/files/data-flow/acra-archi-tls.png)
  
  _Acra-based application architecture using TLS transport encryption._
  
Second way to protect raw data between application and AcraServer/AcraTranslator is to use AcraConnector:

- [_AcraConnector_]({{< ref "/acra/configuring-maintaining/general-configuration/acra_connector.md" >}}) — a client-side daemon that runs under a separate user / in a separate container, and which acts as a database listener that redirects all the queries to AcraServer and, upon receiving the results, feeds it back into the app. AcraConnector is an optional component, required in systems that are using extra transport encryption layer via Themis SecureSession.

  ![](/files/data-flow/acra-entities.png)    
  
  _Acra-based application architecture using TLS transport encryption._
  
### Data flow schemas

With TLS as secure transport between application and AcraServer/AcraTranslator:
1. **Writing**: Application --> AcraTranslator --> Application --> Storage.

   **Description**: application uses AcraTranslator to encrypt/decrypt data and save it in a storage. 
  
   **Reading**: Application --> Storage --> Application --> AcraTranslator --> Application.

>    **Description**: application uses AcraTranslator to encrypt/decrypt data and save it in a storage.

Reviewer: Copy-paste of a description above (line 33).
Maybe make it `application uses AcraTranslator to decrypt data after reading from storage`?
And in line 33 make it `encrypt data` only?

Similar thing is in lines 73 & 78, maybe it was intended...

Author: nice catch. yes, probably copy-paste. applied your suggestion and updated descriptions

2. **Writing**: Application --> AcraWriter --> Database.

   **Description**: application uses AcraWriter and public key to encrypt data and save it directly to database.

   **Reading**: Database --> AcraServer --> Application.

   **Description**: application request data from database through AcraServer that decrypts it transparently.
  
3. **Writing**: Application --> AcraServer --> Database.

   **Description**: application sends data to database through AcraServer in [Transparent encryption mode](/acra/configuring-maintaining/general-configuration/acra-server/#transparent-proxy-mode-INVALID).
   All data manipulations like [encryption]({{< ref "/acra/security-controls/encryption/" >}})/
   [tokenization]({{< ref "/acra/security-controls/tokenisation/" >}})/[masking]({{< ref "/acra/security-controls/masking/" >}})
   applied on AcraServer side before sending to database. 

   **Reading**: Database --> AcraServer --> Application.

   **Description**: application request data from database through AcraServer. AcraServer transparently

>    decrypts/detokinze/unmask data and pass it to application.

Reviewer: ```suggestion
   decrypts/detokinzes/unmasks data and passes it to application.
```

Author: applies it manually because of `detokenizes`, not `detokinzes` ))

   {{< /hint >}}

<figure>
  <img src="/files/data-flow/acra-archi-tls-write-flow.png">
  <figcaption>Data flow #2 and #3</figcaption>
</figure>
  
With SecureSession and AcraConnector for secure transport:

1. **Writing**: Application --> AcraConnector --> AcraTranslator --> AcraConnector --> Application --> Storage.

   **Description**: application uses AcraTranslator to encrypt/decrypt data and save it in a storage. All data sent through
   AcraConnector deployed close to application to protect decrypted data sent from AcraTranslator via SecureSession.

   **Reading**: Application --> Storage --> Application --> AcraConnector --> AcraTranslator --> AcraConnector --> Application.

   **Description**: pplication gets data from storage and decrypts it using AcraTranslator. All data sent through
   AcraConnector deployed close to application to protect decrypted data sent from AcraTranslator via SecureSession.

2. **Writing**: Application --> AcraWriter --> Database.
   
   **Description**: application uses AcraWriter and public key to encrypt data and save it directly to database.

   **Reading**: Database --> AcraServer --> AcraConnector --> Application.

>    **Description**: application request data from database through AcraServer that decrypts it transparently. 

Reviewer: ```suggestion
   **Description**: application requests data from database through AcraServer that decrypts it transparently. 
```

>    AcraConnector with Secure Session protect raw data decrypted from AcraServer sent to application.

Reviewer: ```suggestion
   AcraConnector uses Secure Session to protect raw data decrypted by AcraServer before sending it to application.
```
